crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,593 @@
+        self.demux.insert(cap.clone(), ingress_tx);
+
+        Ok(ingress)
+    }
+
+    fn unmask_msg_id(&self, id: &mut u8) -> Result<&SharedCapability, MuxDemuxError> {
+        for cap in self.shared_capabilities.iter_caps() {
+            let offset = cap.relative_message_id_offset();
+            let next_offset = offset + cap.num_messages();
+            if *id < next_offset {
+                *id -= offset;
+                return Ok(cap)
+            }
+        }
+
+        Err(MessageIdOutOfRange(*id))
+    }
+
+    /// Masks message id with offset relative to the message id suffix reserved for capability
+    /// message ids. The p2p stream further masks the message id (todo: mask whole message id at
+    /// once to avoid copying message to mutate id byte or sink BytesMut).
+    fn mask_msg_id(&self, msg: Bytes) -> Bytes {
+        let mut masked_bytes = BytesMut::zeroed(msg.len());
+        masked_bytes[0] = msg[0] + self.owner.relative_message_id_offset();
+        masked_bytes[1..].copy_from_slice(&msg[1..]);
+
+        masked_bytes.freeze()
+    }
+
+    /// Checks if all clones of this shared stream have been dropped, if true then returns //
+    /// function to drop the stream.
+    fn can_drop(&mut self) -> bool {
+        for tx in self.demux.values() {
+            if !tx.is_closed() {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl<S, E> Stream for MuxDemuxStream<S>
+where
+    S: Stream<Item = Result<BytesMut, E>> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E> + From<<S as Sink<Bytes>>::Error>,
+{
+    type Item = Result<BytesMut, MuxDemuxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut send_count = 0;
+        let mut mux_exhausted = false;
+
+        loop {
+            // send buffered bytes from `StreamClone`s. try send at least as many messages as
+            // there are stream clones.
+            if self.inner.poll_ready_unpin(cx).is_ready() {
+                if let Ok(item) = self.mux.try_recv() {
+                    self.inner.start_send_unpin(item)?;
+                    if send_count < self.demux.len() {
+                        send_count += 1;
+                        continue
+                    }
+                } else {
+                    mux_exhausted = true;
+                }
+            }
+
+            // advances the wire and either yields message for the owner or delegates message to a
+            // stream clone
+            let res = self.inner.poll_next_unpin(cx);
+            if res.is_pending() {
+                // no message is received. continue to send messages from stream clones as long as
+                // there are messages left to send.
+                if !mux_exhausted && self.inner.poll_ready_unpin(cx).is_ready() {
+                    continue
+                }
+                cx.waker().wake_by_ref();
```

Comment:
I don't think we need this, because we then either return or start another iteration, right?
and we call poll_recv above then the channel should have taken care of this